File: devel/libanalysis/src/analysis_module.c
/*
   Copyright (C) 2011  Statoil ASA, Norway. 
    
   The file 'analysis_module.c' is part of ERT - Ensemble based Reservoir Tool. 
    
   ERT is free software: you can redistribute it and/or modify 
   it under the terms of the GNU General Public License as published by 
   the Free Software Foundation, either version 3 of the License, or 
   (at your option) any later version. 
    
   ERT is distributed in the hope that it will be useful, but WITHOUT ANY 
   WARRANTY; without even the implied warranty of MERCHANTABILITY or 
   FITNESS FOR A PARTICULAR PURPOSE.   
    
   See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html> 
   for more details. 
*/

#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <stdio.h>
#include <dlfcn.h>
#include <matrix.h>
#include <util.h>

#include <analysis_module.h>
#include <analysis_table.h>

#define ANALYSIS_MODULE_TYPE_ID 6610123

struct analysis_module_struct {
  UTIL_TYPE_ID_DECLARATION;
  void                       * lib_handle;
  void                       * module_data;

  analysis_free_ftype        * freef;
  analysis_alloc_ftype       * alloc;
  analysis_initX_ftype       * initX;
    
  analysis_set_int_ftype     * set_int;
  analysis_set_double_ftype  * set_double;
  analysis_set_string_ftype  * set_string;
  
  bool                         need_ED;
  bool                         need_randrot;

  char                       * user_name;   /* String used to identify this module for the user; not used in 
                                               the linking process. */
};



static analysis_module_type * analysis_module_alloc_empty( const char * user_name ) {
  analysis_module_type * module = util_malloc( sizeof * module , __func__ );
  UTIL_TYPE_ID_INIT( module , ANALYSIS_MODULE_TYPE_ID );

  module->lib_handle  = NULL;
  module->initX       = NULL;
  module->set_int     = NULL;
  module->set_double  = NULL;
  module->set_string  = NULL;
  module->module_data = NULL;
  module->user_name   = util_alloc_string_copy( user_name );

  return module;
}


static analysis_module_type * analysis_module_alloc__( const analysis_table_type * table , const char * user_name , void * lib_handle ) {
  analysis_module_type * module = analysis_module_alloc_empty( user_name );
  
  module->lib_handle        = lib_handle;
  module->initX             = table->initX;
  module->set_int           = table->set_int;
  module->set_double        = table->set_double;
  module->set_string        = table->set_string;
  module->alloc             = table->alloc;
  module->freef             = table->freef;
  module->need_ED           = table->need_ED;
  module->need_randrot      = table->need_randrot;

  if (module->alloc != NULL)
    module->module_data = module->alloc( );
  
  return module;
}





static analysis_module_type * analysis_module_alloc( const char * libname , const char * user_name , 
                                                     const char * table_name ) {
  analysis_module_type * module = NULL;
  void * lib_handle = dlopen( NULL , RTLD_NOW );
  if (lib_handle != NULL) {
    analysis_table_type * analysis_table = (analysis_table_type *) dlsym( lib_handle , table_name );
    if (analysis_table != NULL) {
      analysis_table_type * analysis  = analysis_table;
      module = analysis_module_alloc__( analysis , user_name , lib_handle );
    } else
      fprintf(stderr , "Failed to load symbol table:%s. Error:%s \n",table_name , dlerror());
    
    if (module == NULL) 
      dlclose( lib_handle );
  } else 
    fprintf(stderr , "Failed to load library:%s. Error:%s \n",libname , dlerror());
  
  return module;
}

analysis_module_type * analysis_module_alloc_internal( const char * user_name , const char * symbol_table ) {
  return analysis_module_alloc( NULL , user_name , symbol_table );
}


analysis_module_type * analysis_module_alloc_external( const char * user_name , const char * lib_name) {
  return analysis_module_alloc( lib_name , user_name , EXTERNAL_MODULE_TABLE );
}



/*****************************************************************/

static UTIL_SAFE_CAST_FUNCTION( analysis_module , ANALYSIS_MODULE_TYPE_ID )


void analysis_module_free( analysis_module_type * module ) {
  if (module->freef != NULL)
    module->freef( module->module_data );
  
  dlclose( module->lib_handle );
  free( module );
}


void analysis_module_free__( void * arg) {
  analysis_module_type * module = analysis_module_safe_cast( arg );
  analysis_module_free( module );
}


void analysis_module_initX(analysis_module_type * module , 
                           matrix_type * X , 
                           matrix_type * S , 
                           matrix_type * R , 
                           matrix_type * innov , 
                           matrix_type * E , 
                           matrix_type * D , 
                           matrix_type * randrot) {
  module->initX(module->module_data , X , S , R , innov , E , D , randrot );
}

bool analysis_module_needs_ED( const analysis_module_type * module ) {
  return module->need_ED;
}

bool analysis_module_needs_randrot( const analysis_module_type * module ) {
  return module->need_randrot;
}

/*****************************************************************/

static bool analysis_module_set_int(analysis_module_type * module , const char * flag , int value) {
  if (module->set_int != NULL) 
    return module->set_int( module->module_data , flag , value );
  else
    return false;
}

static bool analysis_module_set_double(analysis_module_type * module , const char * var , double value) {
  if (module->set_double != NULL)
    return module->set_double( module->module_data , var , value );
  else
    return false;
}

static bool analysis_module_set_string(analysis_module_type * module , const char * var , const char * value) {
  if (module->set_string != NULL)
    return module->set_string( module->module_data , var , value );
  else
    return false;
}

/* 
   The input value typically comes from the ocnfiguration system and
   is in terms of a string - irrespective of the fundamental type of
   the underlying parameter. The algorithm for setting the parameter
   tries datatypes as follows: integer - double - string.

   For the numeric datatypes the algorithm is two step:

     1. Try the conversion string -> numeric.
     2. Try calling the analysis_module_set_xxx() function.

   Observe that this implies that the same variable name can NOT be
   used for different variable types.
*/

bool analysis_module_set_var( analysis_module_type * module , const char * var_name , const char * string_value ) {
  bool set_ok = false;
  {
    int  int_value;
    
    if (util_sscanf_int( string_value , &int_value )) 
      set_ok = analysis_module_set_int( module , var_name , int_value );
    
    if (set_ok)
      return true;
  }
  
  {
    double double_value;
    if (util_sscanf_double( string_value , &double_value )) 
      set_ok = analysis_module_set_double( module , var_name , double_value );

    if (set_ok)
      return true;
  }
  set_ok = analysis_module_set_string( module , var_name , string_value );
  if (!set_ok)
    fprintf(stderr,"** Warning: failed to set %s=%s for analysis module:%s\n", var_name , string_value , module->user_name);
  
  return set_ok;
}

File: python/res/analysis/analysis_module.py
#  Copyright (C) 2013  Equinor ASA, Norway.
#
#  The file 'analysis_module.py' is part of ERT - Ensemble based Reservoir Tool.
#
#  ERT is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  ERT is distributed in the hope that it will be useful, but WITHOUT ANY
#  WARRANTY; without even the implied warranty of MERCHANTABILITY or
#  FITNESS FOR A PARTICULAR PURPOSE.
#
#  See the GNU General Public License at <http://www.gnu.org/licenses/gpl.html>
#  for more details.

from cwrap import BaseCClass
from ecl.util.util.rng import RandomNumberGenerator
from res import ResPrototype
from os import path

import res
from res.util import Matrix


class AnalysisModule(BaseCClass):
    TYPE_NAME = "analysis_module"

    _alloc_external = ResPrototype(
        "void* analysis_module_alloc_external(char*)", bind=False
    )
    _alloc_internal = ResPrototype(
        "void* analysis_module_alloc_internal(char*)", bind=False
    )
    _free = ResPrototype("void analysis_module_free(analysis_module)")
    _get_lib_name = ResPrototype("char* analysis_module_get_lib_name(analysis_module)")
    _get_module_internal = ResPrototype(
        "bool analysis_module_internal(analysis_module)"
    )
    _set_var = ResPrototype(
        "bool analysis_module_set_var(analysis_module, char*, char*)"
    )
    _get_table_name = ResPrototype(
        "char* analysis_module_get_table_name(analysis_module)"
    )
    _get_name = ResPrototype("char* analysis_module_get_name(analysis_module)")
    _check_option = ResPrototype(
        "bool analysis_module_check_option(analysis_module, analysis_module_options_enum)"
    )
    _has_var = ResPrototype("bool analysis_module_has_var(analysis_module, char*)")
    _get_double = ResPrototype(
        "double analysis_module_get_double(analysis_module, char*)"
    )
    _get_int = ResPrototype("int analysis_module_get_int(analysis_module, char*)")
    _get_bool = ResPrototype("bool analysis_module_get_bool(analysis_module, char*)")
    _get_str = ResPrototype("char* analysis_module_get_ptr(analysis_module, char*)")
    _init_update = ResPrototype(
        "void analysis_module_init_update(analysis_module, bool_vector, bool_vector, matrix, matrix, matrix, matrix, matrix, rng)"
    )
    _updateA = ResPrototype(
        "void analysis_module_updateA(analysis_module, matrix, matrix, matrix, matrix, matrix, matrix, void*, rng)"
    )
    _initX = ResPrototype(
        "void analysis_module_initX(analysis_module, matrix, matrix, matrix, matrix, matrix, matrix, matrix, rng)"
    )

    # The VARIABLE_NAMES field is a completly broken special case
    # which only applies to the rml module.
    VARIABLE_NAMES = {
        "IES_MAX_STEPLENGTH": {
            "type": float,
            "description": "Max step Length of Gauss Newton Iteration",
        },
        "IES_MIN_STEPLENGTH": {
            "type": float,
            "description": "Min step Length of Gauss Newton Iteration",
        },
        "IES_DEC_STEPLENGTH": {
            "type": float,
            "description": "Decline of step Length in Gauss Newton Iteration",
        },
        "IES_INVERSION": {"type": int, "description": "Inversion algorithm"},
        "IES_DEBUG": {
            "type": bool,
            "description": "Print extensive log for IES analysis steps",
        },
        "IES_LOGFILE": {"type": str, "description": "IES Log File"},
        "IES_AAPROJECTION": {
            "type": str,
            "description": "Include projection Y (A^+A) for n<N-1",
        },
        "LAMBDA0": {"type": float, "description": "Initial Lambda"},
        "USE_PRIOR": {
            "type": bool,
            "description": "Use both Prior and Observation Variability",
        },
        "LAMBDA_REDUCE": {"type": float, "description": "Lambda Reduction Factor"},
        "LAMBDA_INCREASE": {"type": float, "description": "Lambda Incremental Factor"},
        "LAMBDA_MIN": {"type": float, "description": "Minimum Lambda"},
        "LOG_FILE": {"type": str, "description": "Log File"},
        "CLEAR_LOG": {"type": bool, "description": "Clear Existing Log File"},
        "LAMBDA_RECALCULATE": {
            "type": bool,
            "description": "Recalculate Lambda after each Iteration",
        },
        "ENKF_TRUNCATION": {"type": float, "description": "Singular value truncation"},
        "ENKF_SUBSPACE_DIMENSION": {
            "type": int,
            "description": "Number of singular values",
        },
        "ENKF_NCOMP": {"type": int, "description": "Number of singular values"},
        "CV_NFOLDS": {"type": int, "description": "CV_NFOLDS"},
        "FWD_STEP_R2_LIMIT": {"type": float, "description": "FWD_STEP_R2_LIMIT"},
        "CV_PEN_PRESS": {"type": bool, "description": "CV_PEN_PRESS"},
    }

    def __init__(self, name=None, lib_name=None):
        if name is None and lib_name is None:
            raise ValueError("Must supply exactly one of lib or lib_name")

        if name and lib_name:
            raise ValueError("Must supply exactly one of name or lib_name")

        if lib_name:
            c_ptr = self._alloc_external(lib_name)
        else:
            c_ptr = self._alloc_internal(name)
            if not c_ptr:
                raise KeyError("Failed to load internal module:%s" % name)

        super(AnalysisModule, self).__init__(c_ptr)

    def getVariableNames(self):
        """ @rtype: list of str """
        items = []
        for name in AnalysisModule.VARIABLE_NAMES:
            if self.hasVar(name):
                items.append(name)
        return items

    def getVariableValue(self, name):
        """ @rtype: int or float or bool or str """
        variable_type = self.getVariableType(name)
        if variable_type == float:
            return self.getDouble(name)
        elif variable_type == bool:
            return self.getBool(name)
        elif variable_type == str:
            return self.getStr(name)
        elif variable_type == int:
            return self.getInt(name)

    def getVariableType(self, name):
        """ :rtype: type """
        return AnalysisModule.VARIABLE_NAMES[name]["type"]

    def getVariableDescription(self, name):
        """ :rtype: str """
        return AnalysisModule.VARIABLE_NAMES[name]["description"]

    def getVar(self, name):
        return self.getVariableValue(name)

    def free(self):
        self._free()

    def __repr__(self):
        if not self:
            return repr(None)
        nm = self.name()
        tn = self.getTableName()
        ln = self.getLibName()
        mi = "internal" if self.getInternal() else "external"
        ad = self._ad_str()
        fmt = "AnalysisModule(name = %s, table = %s, lib = %s, %s) %s"
        return fmt % (nm, tn, ln, mi, ad)

    def getLibName(self):
        return self._get_lib_name()

    def getInternal(self):
        return self._get_module_internal()

    def __assertVar(self, var_name):
        if not self.hasVar(var_name):
            raise KeyError("Module does not support key:%s" % var_name)

    def setVar(self, var_name, value):
        self.__assertVar(var_name)
        string_value = str(value)
        return self._set_var(var_name, string_value)

    def getTableName(self):
        return self._get_table_name()

    def getName(self):
        """ :rtype: str """
        return self.name()

    def name(self):
        return self._get_name()

    def checkOption(self, flag):
        return self._check_option(flag)

    def hasVar(self, var):
        """ :rtype: bool """
        return self._has_var(var)

    def getDouble(self, var):
        """ :rtype: float """
        self.__assertVar(var)
        return self._get_double(var)

    def getInt(self, var):
        """ :rtype: int """
        self.__assertVar(var)
        return self._get_int(var)

    def getBool(self, var):
        """ :rtype: bool """
        self.__assertVar(var)
        return self._get_bool(var)

    def getStr(self, var):
        """ :rtype: str """
        self.__assertVar(var)
        return self._get_str(var)

    def initUpdate(self, ens_mask, obs_mask, S, R, dObs, E, D, rng):
        self._init_update(ens_mask, obs_mask, S, R, dObs, E, D, rng)

    def updateA(self, A, S, R, dObs, E, D, rng):
        self._updateA(A, S, R, dObs, E, D, None, rng)

    def initX(self, A, S, R, dObs, E, D, rng):
        X = Matrix(A.columns(), A.columns())
        self._initX(X, A, S, R, dObs, E, D, rng)
        return X

    def __ne__(self, other):
        """
        not equal operator between two modules
        :param other: other module to compare with
        :return: True if different
        """
        return not self == other

    def __eq__(self, other):
        """
        equality operator between two modules
        :param other: other module to compare with
        :return: True if the same
        """

        if self.getName() != other.getName():
            return False
        if self.getLibName() != other.getLibName():
            return False
        if self.getTableName() != other.getTableName():
            return False
        if self.getInternal() != other.getInternal():
            return False

        var_name_local = self.getVariableNames()
        var_name_other = other.getVariableNames()

        if var_name_local != var_name_other:
            return False

        for a in var_name_local:
            if self.getVariableValue(a) != other.getVariableValue(a):
                return False

        return True
